packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js

```diff
@@ -0,0 +1,166 @@
+ * @param {string} [params.providerCumulusId]
+ *   Cumulus ID for provider referenced in workflow message, if any
+ * @param {string} [params.pdrCumulusId]
+ *   Cumulus ID for PDR referenced in workflow message, if any
+ * @param {Object} [params.granuleModel]
+ *   Optional override for the granule model writing to DynamoDB
+ *
+ * @returns {Promise}
+ * @throws
+ */
+const writeGranule = async ({
+  granule,
+  cumulusMessage,
+  executionDescription,
+  executionUrl,
+  collectionCumulusId,
+  executionCumulusId,
+  knex,
+  providerCumulusId,
+  pdrCumulusId,
+  granuleModel,
+}) =>
+  knex.transaction(async (trx) => {
+    await writeGranuleViaTransaction({
+      cumulusMessage,
+      granule,
+      collectionCumulusId,
+      providerCumulusId,
+      executionCumulusId,
+      pdrCumulusId,
+      trx,
+    });
+    return granuleModel.storeGranuleFromCumulusMessage({
+      granule,
+      cumulusMessage,
+      executionUrl,
+      executionDescription,
+    });
+  });
+
+/**
+ * Write granules to DynamoDB and Postgres
+ *
+ * @param {Object} params
+ * @param {Object} params.cumulusMessage - A workflow message
+ * @param {string} params.collectionCumulusId
+ *   Cumulus ID for collection referenced in workflow message, if any
+ * @param {string} params.executionCumulusId
+ *   Cumulus ID for execution referenced in workflow message, if any
+ * @param {Knex} params.knex - Client to interact with Postgres database
+ * @param {string} [params.providerCumulusId]
+ *   Cumulus ID for provider referenced in workflow message, if any
+ * @param {string} [params.pdrCumulusId]
+ *   Cumulus ID for PDR referenced in workflow message, if any
+ * @param {Object} [params.granuleModel]
+ *   Optional override for the granule model writing to DynamoDB
+ *
+ * @returns {Promise<Object[]>}
+ *  true if there are no granules on the message, otherwise
+ *  results from Promise.allSettled for all granules
+ * @throws {Error} - if no collection is provided
+ */
+const writeGranules = async ({
+  cumulusMessage,
+  collectionCumulusId,
+  executionCumulusId,
+  knex,
+  providerCumulusId,
+  pdrCumulusId,
+  granuleModel = new Granule(),
+}) => {
+  if (!collectionCumulusId) {
+    throw new Error('Collection reference is required for granules');
+  }
+
+  const granules = getMessageGranules(cumulusMessage);
+  const executionArn = getMessageExecutionArn(cumulusMessage);
+  const executionUrl = getExecutionUrlFromArn(executionArn);
+
+  let executionDescription;
+  try {
+    executionDescription = await describeExecution({ executionArn });
+  } catch (error) {
+    log.error(`Could not describe execution ${executionArn}`, error);
+  }
+
+  // Process each granule in a separate transaction via Promise.allSettled
+  // so that they can succeed/fail independently
+  const results = await Promise.allSettled(granules.map(
+    (granule) => writeGranule({
+      granule,
+      cumulusMessage,
+      executionDescription,
```

Comment:
This signature was updated to require executionDescription, executionUrl to move the calls to getMessageExecutionArn getExecutionUrlFromArn to the parent method.    👍

Reply:
Yeah, I realized that it was foolish to do the call to get the execution **for each granule**, so I refactored it so that the call to the describe the execution only happens once per lambda invocation in writeGranules